### PR TITLE
feat: tictactoe example application

### LINKED ISSUE #109: TicTacToe smart contract
# Summary

The game is simple in its original state which consists of

- 3x3 grid
- 2 players
- Each player makes a move in a turn based fashion
- First player to reach 3 in a row wins otherwise it's a draw

Expansions to game logic (more players, larger boards, more / less markers required to win etc.) and contract data handling (multiple games at once, allowing contracts to interact with the game etc.) can be added later.

## Rough Contract Design

- The contract should support 2 players and they should not be contracts
    - The creator of the game can specify themeselves as both players
- After a player makes a move the next move is automatically set to be made by the next player unless the game has ended
- The game ends in any of the following states
    - Either player has reached 3 in a row
        - This can be in any direction (horizontal, vertical or diagonal)
    - The grid is full and no more moves can be made
- A player cannot place their marker in a position that is already filled
- A player cannot make a move outside of the designated 3x3 grid
- There can only be 1 game played at a time
    - This means that the game itself is stored in storage
    - No other storage is required
- Starting a new game overwrites any previous game even if that game has not ended
- Events should be emitted when
    - A new game is created
        - Who the players are
    - A move is made
        - Which player has made a move and at which position
    - The game ends (draw or if someone won then who won)

### PR DESCRIPTION
**Pull request by djma**
## Type of change

- New feature

## Changes

The following changes have been made:

- Introduce a new application: tic-tac-toe

## Notes

- Currently stuck at with out-of-gas failure. I'm not sure how to provide more gas to the contract call.
```
failures:

---- functions::play_move::revert::move_already_played stdout ----
thread 'functions::play_move::revert::move_already_played' panicked at 'called `Result::unwrap()` on an `Err` value: ProviderError("gas_limit(1000000) is lower than the estimated gas_used(1002095)")', project/contracts/tictactoe-contract/tests/utils/mod.rs:97:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: panic did not contain expected string
      panic message: `"called `Result::unwrap()` on an `Err` value: ProviderError(\"gas_limit(1000000) is lower than the estimated gas_used(1002095)\")"`,
 expected substring: `"Revert(18446744073709486080)"`
```

- I wanted to refactor the game state calculation into a function but the compiler complained about all calls to `storage` having to be in `main.sw`. I tried a few things but ended up giving up.

## Related Issues

Closes #109

### CODE REVIEW

**Comment by Braqzen** [comment]
cc @Rubyt0x

**Comment by sezna** [comment]
CI is looking correct now 😄

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -40,7 +40,8 @@ jobs:
 
     strategy:
       matrix:
-        project: ["airdrop", "AMM", "auctions", "DAO", "escrow", "fundraiser", "multisig-wallet", "name-registry", "NFT", "oracle", "OTC-swap-predicate"]
+        project: ["airdrop", "AMM", "auctions", "DAO", "escrow", "fundraiser", "multisig-wallet", "name-registry", "NFT", "oracle", "OTC-swap-predicate",
+        "tictactoe"]
```
Comment (sezna) [inline .github/workflows/ci.yml:44]:
```suggestion
        project: ["airdrop", "AMM", "auctions", "DAO", "escrow", "fundraiser", "multisig-wallet", "name-registry", "NFT", "oracle", "OTC-swap-predicate", "tictactoe"]
```

**Review by sezna — COMMENTED** [review]
Just need to run `forc fmt` in the project root dir to fmt the sway files and CI should pass!

**Comment by Braqzen** [comment]
Closing because

1) Stale
2) Another contributor has submitted a PR which closed the issue #157

**Pull request closed without merging**